In [42]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np


# Read full dataset
df = pd.read_csv('RideTrip.csv', low_memory = False)
# Set display options for better visibility
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

# Print head (first 10 rows)
print("First 10 rows:")
df.head(10)

First 10 rows:


,Started On Date,Started On Time,Completed On Date,Completed On Time,Distance Travelled,End Location Lat,End Location Long,Driver Rating,Rider Rating,Start Zip Code,End Zip Code,Requested Car Category,Surge Factor,Start Location Long,Start Location Lat,Color,Make,Model,Year,Rating,PRCP,TMAX,TMIN,AWND,GustSpeed2,Fog,HeavyFog,Thunder,ID,Holiday,Weekend,cancelled,Number of rides that day
0,2016-06-04,08:34:52,2016-06-04,08:35:56,285.0,30.27,-97.75,5.0,5,-1,-1,REGULAR,0.0,-97.75,30.27,Black,Cadillac,XTS,2013,5.000000,0.1,86.000000,67.000000,4.9,13.000000,1,0,0,2399,0,1,1,1
1,2016-06-04,08:45:01,2016-06-04,08:51:15,1029.0,30.27,-97.74,5.0,5,-1,-1,REGULAR,0.0,-97.75,30.27,Black,Cadillac,XTS,2013,5.000000,0.1,86.000000,67.000000,4.9,13.000000,1,0,0,2399,0,1,0,2
2,2016-06-04,09:18:49,2016-06-04,09:27:32,8459.0,38.68,-121.04,5.0,5,-1,-1,REGULAR,0.0,-121.07,38.65,Gray,Bentley,Continental GT,2013,5.000000,0.1,86.000000,67.000000,NaN,13.000000,1,0,0,1557,0,1,0,1
3,2016-06-04,10:50:12,2016-06-04,10:51:49,443.0,38.68,-121.04,5.0,5,-1,-1,REGULAR,0.0,-121.04,38.68,Gray,Bentley,Continental GT,2013,5.000000,0.1,86.000000,67.545955,4.9,12.540277,1,0,0,1557,0,1,1,2
4,2016-06-04,12:16:02,2016-06-04,12:17:57,568.0,38.68,-121.04,3.0,5,-1,-1,REGULAR,0.0,-121.04,38.68,Gray,Bentley,Continental GT,2013,5.000000,0.1,85.620379,67.000000,4.9,13.000000,1,0,0,1557,0,1,0,3
5,2016-06-04,19:05:13,2016-06-04,19:13:38,4051.0,30.27,-97.74,5.0,5,-1,-1,REGULAR,0.0,-97.76,30.25,Black,Cadillac,XTS,2013,5.000000,0.1,86.000000,67.000000,4.9,13.000000,1,0,0,2399,0,1,0,3
6,2016-06-04,19:24:25,2016-06-04,19:26:07,790.0,30.27,-97.75,5.0,5,-1,-1,REGULAR,0.0,-97.75,30.27,Black,Cadillac,XTS,2013,5.000000,0.1,-41.109399,67.000000,4.9,13.000000,1,0,0,2399,0,1,0,4
7,2016-06-05,08:17:40,2016-06-05,08:33:20,10260.0,30.27,-97.75,-1.0,5,-1,-1,REGULAR,0.0,-97.77,30.20,Black,Infiniti,QX60,2015,5.000000,0.0,88.000000,68.000000,5.8,14.100000,0,0,0,1178,0,1,0,1
8,2016-06-05,10:57:45,2016-06-05,11:12:48,5294.0,30.24,-97.78,3.0,5,-1,-1,REGULAR,0.0,-97.75,30.27,Black,Cadillac,SRX,2012,4.888889,NaN,88.000000,68.000000,5.8,14.100000,0,0,0,1404,0,1,0,1
9,2016-06-05,11:26:06,2016-06-05,11:36:36,9768.0,30.20,-97.77,4.5,5,-1,-1,REGULAR,0.0,-97.74,30.27,Silver,Toyota,Highlander Hybrid,2008,5.000000,NaN,88.224805,68.000000,5.8,14.100000,0,0,0,1669,0,1,0,1


In [43]:
# Print tail (last 10 rows)
print("Last 10 rows:")
df.tail(10)

Last 10 rows:


,Started On Date,Started On Time,Completed On Date,Completed On Time,Distance Travelled,End Location Lat,End Location Long,Driver Rating,Rider Rating,Start Zip Code,End Zip Code,Requested Car Category,Surge Factor,Start Location Long,Start Location Lat,Color,Make,Model,Year,Rating,PRCP,TMAX,TMIN,AWND,GustSpeed2,Fog,HeavyFog,Thunder,ID,Holiday,Weekend,cancelled,Number of rides that day
862917,2017-02-07,13:20:09,2017-02-07,13:30:21,6131.0,30.28,-97.74,-1.0,5,78752,78705,REGULAR,1.0,-97.72,30.33,Silver,Nissan,Altima,2013,4.977778,0.0,86.000000,-46.654181,2.200000,13.0,1,0,0,2302,0,0,0,4
862918,2017-02-07,13:23:06,2017-02-07,13:33:03,7998.0,30.24,-97.72,-1.0,5,78705,78741,REGULAR,1.0,-97.74,30.28,Black,Audi,SQ5,2016,4.955556,0.0,86.000000,56.000000,2.200000,13.0,1,0,0,363,0,0,0,1
862919,2017-02-07,13:25:16,2017-02-07,13:42:42,10487.0,30.24,-97.71,-1.0,5,78705,78741,REGULAR,1.0,-97.74,30.29,Grey,Nissan,Versa,2016,5.000000,0.0,86.000000,56.000000,2.200000,13.0,1,0,0,93,0,0,0,6
862920,2017-02-07,13:35:43,2017-02-07,13:43:38,6256.0,30.27,-97.69,-1.0,5,78701,78721,REGULAR,1.0,-97.74,30.27,Gray,Chevrolet,Colorado,2016,5.000000,0.0,86.000000,56.000000,2.200000,13.0,1,0,0,652,0,0,0,3
862921,2017-02-07,13:35:50,2017-02-07,13:46:21,8558.0,30.29,-97.75,-1.0,5,78723,78705,REGULAR,1.0,-97.71,30.31,White,Toyota,Camry,2011,5.000000,0.0,86.000000,56.000000,2.200000,13.0,1,0,0,1845,0,0,0,17
862922,2017-02-07,13:39:10,2017-02-07,13:44:48,4272.0,30.29,-97.72,-1.0,5,78701,78722,REGULAR,1.0,-97.75,30.27,Gray,Mazda,Tribute,2008,5.000000,0.0,86.000000,56.000000,2.200000,13.0,1,0,0,1115,0,0,0,6
862923,2017-02-07,13:45:32,2017-02-07,14:00:09,16987.0,30.18,-97.81,5.0,5,78701,78748,REGULAR,1.0,-97.74,30.27,Silver,Subaru,XV Crosstrek,2013,5.000000,0.0,176.214556,56.000000,2.200000,13.0,1,0,0,339,0,0,0,4
862924,2017-02-07,13:47:16,2017-02-07,14:02:11,17818.0,30.20,-97.67,-1.0,5,78704,78719,REGULAR,1.0,-97.79,30.25,Black,Subaru,Outback Sport,2009,5.000000,0.0,86.000000,56.000000,2.200000,13.0,1,0,0,1741,0,0,0,1
862925,2017-02-07,13:54:05,2017-02-07,14:11:37,20007.0,30.40,-97.67,-1.0,4,78701,78753,REGULAR,1.0,-97.75,30.27,Silver,Toyota,Corolla,2006,5.000000,0.0,86.000000,56.000000,2.200000,13.0,1,0,0,277,0,0,0,1
862926,2017-02-07,14:09:22,2017-02-07,14:19:12,6732.0,30.22,-97.74,-1.0,5,78702,78741,REGULAR,1.0,-97.71,30.26,Silver,Honda,Accord,2013,5.000000,0.0,86.000000,56.000000,2.388035,13.0,1,0,0,270,0,0,0,2


### Initial Data Inspection: First and Last 5 Rows

We used `df.head()` and `df.tail()` to quickly check the structure of the dataset.

**What we see:**
- Dataset has 862,927 data entries.
- The dataset has 33 columns and includes trip details (distance, location), driver/rider info, weather data, and the target variable `cancelled` (0 = completed, 1 = cancelled).
- Some columns have missing or placeholder values:
    - `-1` appears in `Start Zip Code`, `End Zip Code`, and `Driver Rating`
    - empty cells are `NaN` 
- The `cancelled` column shows both 0s and 1s in the sample, so we have data on both completed and cancelled rides to analyze.
- Date & time columns (`Started On Date`, `Started On Time`, etc.) are stored as strings and will need to be converted to datetime format for feature engineering.


So missing values marked as `-1` or `NaN` need to be handled and we need to extract useful features from the datetime columns (e.g., hour of day, day of week, duration).

In [36]:
# Dataset shape
print("Dataset shape:", df.shape)

Dataset shape: (862927, 33)


### Raw Dataset Dimensions & Structure
The dataset has 862,927 rows and 33 columns (features + target).

In [53]:
# Data types and non-null counts
print("Column info:")
df.info()

Column info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 862927 entries, 0 to 862926
Data columns (total 33 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Started On Date           862927 non-null  object 
 1   Started On Time           862927 non-null  object 
 2   Completed On Date         862927 non-null  object 
 3   Completed On Time         862927 non-null  object 
 4   Distance Travelled        862927 non-null  float64
 5   End Location Lat          862927 non-null  float64
 6   End Location Long         862927 non-null  float64
 7   Driver Rating             862927 non-null  float64
 8   Rider Rating              862927 non-null  int64  
 9   Start Zip Code            862927 non-null  object 
 10  End Zip Code              862927 non-null  object 
 11  Requested Car Category    862927 non-null  object 
 12  Surge Factor              862927 non-null  float64
 13  Start Location Long       86292

- All 33 columns have 862,927 non-null entries except weather columns (`PRCP`, `TMAX`, `TMIN`, `AWND`, `GustSpeed2`), which have slightly fewer (~837,000). This is mostly weather data, which means roughly 3% of weather data is missing.

- Data types:
  - `object`: Date/time columns, zip codes, car category, color, make, model. These will need encoding later.
  - `float64`: Numeric features like distance, location, ratings, weather.
  - `int64`: Binary & integer features like `cancelled`, `Holiday`, `Weekend`.

So missing weather data will need imputation before modeling.

Categorical columns will need to be converted to numeric format through encoding.

The target `cancelled` is already integer (0/1), which is ready for classification.

In [40]:
# Summary statistics
print("Summary statistics:")
df.describe()

Summary statistics:


,Distance Travelled,End Location Lat,End Location Long,Driver Rating,Rider Rating,Surge Factor,Start Location Long,Start Location Lat,Year,Rating,PRCP,TMAX,TMIN,AWND,GustSpeed2,Fog,HeavyFog,Thunder,ID,Holiday,Weekend,cancelled,Number of rides that day
count,8.629270e+05,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000,837306.000000,837293.000000,837345.000000,837199.000000,837297.000000,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000,862927.000000
mean,8.806241e+03,30.275470,-97.740867,4.141877,4.762907,1.051985,-97.742286,30.275972,2011.841763,4.975342,0.083463,75.813431,55.421722,4.141354,12.111502,0.419077,0.068037,0.083673,870.470149,0.048780,0.490781,0.012489,6.773350
std,6.501929e+04,0.097903,0.274842,1.965777,1.032813,0.424730,0.138644,0.059982,3.974268,0.067723,0.316521,18.079639,16.905381,2.680934,4.526845,0.493408,0.251810,0.276897,707.905930,0.215409,0.499915,0.111054,6.580752
min,-1.000000e+00,0.000000,-121.040000,-1.000000,-1.000000,0.000000,-121.070000,29.630000,2001.000000,3.111111,-2.426783,-67.826872,-78.525364,-17.130059,-23.791361,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000
25%,3.170000e+03,30.250000,-97.750000,5.000000,5.000000,1.000000,-97.750000,30.250000,2010.000000,4.977778,0.000000,67.000000,47.000000,2.500000,8.900000,0.000000,0.000000,0.000000,285.000000,0.000000,0.000000,0.000000,2.000000
50%,5.815000e+03,30.270000,-97.740000,5.000000,5.000000,1.000000,-97.740000,30.270000,2013.000000,5.000000,0.000000,77.000000,55.000000,3.800000,12.100000,0.000000,0.000000,0.000000,707.000000,0.000000,0.000000,0.000000,5.000000
75%,1.175150e+04,30.290000,-97.730000,5.000000,5.000000,1.000000,-97.730000,30.290000,2015.000000,5.000000,0.000000,86.000000,65.000000,5.400000,14.100000,1.000000,0.000000,0.000000,1276.000000,0.000000,1.000000,0.000000,9.000000
max,4.302576e+07,50.020000,19.900000,5.000000,5.000000,6.000000,19.910000,50.030000,2104.000000,5.000000,2.593893,219.422456,189.390422,25.413662,48.015252,1.000000,1.000000,1.000000,3791.000000,1.000000,1.000000,1.000000,92.000000


- `Distance Travelled`: Min value is -1.0, which is not physically possible. So `-1` is used as a placeholder for missing data.
- `Driver Rating`: Min is -1.0, median is 5.0. Again, `-1` likely means it is missing or no rating recorded.
- `Surge Factor`: Mostly 1.0 (no surge), max 6.0. High surge entries are rare.
- Weather columns (`PRCP`, `TMAX`, etc.): Show some extreme values (e.g., TMAX min = -67.8) that seem unrealistic. These may be data errors.
- `cancelled`: Mean = 0.0125 → only 1.25% of rides are cancelled. This is a highly imbalanced target.

The low cancellation rate means we must use appropriate metrics (precision, recall, F1) and possibly handle class imbalance during modeling.

In [54]:
# Count missing values per column
print(df.isnull().sum())

Started On Date                 0
Started On Time                 0
Completed On Date               0
Completed On Time               0
Distance Travelled              0
End Location Lat                0
End Location Long               0
Driver Rating                   0
Rider Rating                    0
Start Zip Code                  0
End Zip Code                    0
Requested Car Category          0
Surge Factor                    0
Start Location Long             0
Start Location Lat              0
Color                           0
Make                            0
Model                           0
Year                            0
Rating                          0
PRCP                        25621
TMAX                        25634
TMIN                        25582
AWND                        25728
GustSpeed2                  25630
Fog                             0
HeavyFog                        0
Thunder                         0
ID                              0
Holiday       

Interpretation:
Most columns are complete as they have 0 missing values.
Weather columns have significant missing data (roughly 25.6k each in PRCP, TMAX, TMIN, AWND, GustSpeed2).
This suggests weather data may only be available for some of the rides/dates, or was collected inconsistently.
Since these are numeric features, we can consider either: 
- imputation (median/mean)
- creating a "weather_missing" indicator flag
- dropping them if they add little predictive value